packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java

```diff
@@ -85,7 +88,7 @@ final class VideoPlayer {
 
     MediaSource mediaSource = buildMediaSource(uri, dataSourceFactory, formatHint);
 
-    exoPlayer.setMediaSource(mediaSource);
+    unstableSetMediaSource(exoPlayer, mediaSource);
```

Comment:
Do we actually need to use `MediaSource` at all? It looks like we're doing two things with the media source:
1. Setting the source type.
2. Injecting the data source factory.

From the APIs, it looks like 1. can be done with [`MediaItem` MIME types](https://developer.android.com/media/media3/exoplayer/media-items#handling-non-standard) and 2 can be done [as part of `ExoPlayer` creation](https://developer.android.com/media/media3/exoplayer/media-sources#customizing-media-source-creation). Are there reasons those wouldn't work?

Reply:
Well, concrete MediaSource.Factory implementations (which are to be set to builder) are UnstableApi as well

so instead of unstable player.setMediaSource we'd have unstable createSomeFactory

Reply:
This is however one unstable call less
I'll push a commit so you can see whether it's ok

Reply:
Why do we need to create the concrete factories. Don't we just pass in `dataSourceFactory`, and then the internal implementation of ExoPlayer creates the concrete factories from types as needed when it internally creates `MediaSource`s from the `MediaItem`s we pass it? My impression from the docs is that `unstableBuildMediaSource` is largely duplicating logic that ExoPlayer already has.

Reply:
Oh, wasn't sure DefaultMediaSourceFactory is enough, need to check it

Reply:
ExoPlayer dev here chiming in :)

I think the `MediaItem` way of writing this code (without `MediaSource`) would be more like:

```
ExoPlayer exoPlayer = new ExoPlayer.Builder(context)
    .setMediaSourceFactory(new DefaultMediaSourceFactory(dataSourceFactory))
    .build();
    
mediaItem = new MediaItem.Builder
    .setUri(uri)
    .setMimeType(mimeFromFormatHint(formatHint))
    .build();

exoPlayer.setMediaItem(mediaItem);
```

`mimeTypeFromFormatHint` would just map from your own constants to those in [`MimeTypes`](https://developer.android.com/reference/androidx/media3/common/MimeTypes).

See also https://developer.android.com/media/media3/exoplayer/media-items and some examples at https://developer.android.com/media/media3/exoplayer/customization too.

Reply:
@icbaker Thanks for the confirmation!

Reply:
thanks! updated the branch